_posts/2016-03-28-Mutual-Problems.md

```diff
@@ -0,0 +1,130 @@
+
+## Java - the problem, the solution and why is it so hard
+
+Because it is quite unusual to use the mutual authentication every day, together with my teammate we asked for the assistance. The first look at the results served by *uncle Google* didn't revealed the complexity behind the implementation, but each click on the results led us to the more and more confusing solutions (some of them where from 90's). To make matters worse we had to use [Apache HttpComponents](https://hc.apache.org/) to implement our connection, but the most of the proposed solutions based on the pure **Java** libraries.  
+
+The knowledge from the internet allows us to establish:
+
+- **Java** cannot use directly any certificates or private keys (like e.g. `curl`)
+- **Java** requires separate files (**Java Keystores**) which can contain original certificates and keys. 
+- We needed a trusted keystore with the certificate required for the server's certificate verification.
+- We needed a keys keystore with the client's certificate and the client's private key.
+
+First we had to create the trusted keystore. We created the keystore with the certificate using the `keytool` command:
+ 
+    $ keytool -import -alias trusted_certificate -keystore trusted.jks -file trusted.crt
+
+We stored in the keystore file `trusted.jks` the certificate `trusted.crt` under the alias `trusted_certificate`. During the execution of this command we were asked to input a password for this keystore. This password we used later to get access to the keystore file.
+
+To create a keystore a few additional steps were required. In most cases you will probably receive two files from the company which issue the client's certificate. The first file will be the client's certificate in the `pem` format. This certificate will be send to the server. The second file is the client's private key (also in the `pem` format) which is used during the handshake to confirm that you are owner of the client's certificate.
+ 
+Unfortunately **Java** supports only the `PKCS12` format. So we had to translate our certificate and private key to `PKCS12` format. We can do that using OpenSSL.
+
+    $ openssl pkcs12 -export \
+        -in client.crt \
+        -inkey client.key \
+        -out key.p12 \
+        -name client \
+        -CAfile ca.crt \
+        -caname root
+
+We generated the file `key.p12` from files `client.crt`, `client.key` and `ca.crt`. Once again a password input was required. This password is used to protect the private key. 
+This approach worked for us, only because we where the issuer of the client's certificate and we had access to our CA certificate used for signing the client's certificate. If you receive the client's certificate from owners of the server, you can ask them for a certificate and a private key in `PKCS12` format.   
```

Comment:
I didn't check why the CA is required and I don't know why . I will check it.

I deliberately omit a self-signed certificate, because you should not use it on prod env.